File: Java-Core/src/main/java/presto/value/ClosureValue.java
package presto.value;

import presto.declaration.IMethodDeclaration;
import presto.error.PrestoError;
import presto.grammar.ArgumentList;
import presto.grammar.Identifier;
import presto.runtime.Context;
import presto.type.IType;
import presto.type.MethodType;

public class ClosureValue extends BaseValue {
	
	IMethodDeclaration method;
	
	public ClosureValue(Context context, IMethodDeclaration method) {
		super(new MethodType(context, method.getIdentifier()));
		this.method = method;
	}
	
	public IMethodDeclaration getMethodDeclaration() {
		return method;
	}
	
	public IValue interpret(Context context) throws PrestoError {
		Context thisContext = ((MethodType)type).getContext();
		Context parentMost = thisContext.getParentMostContext();
		parentMost.setParentContext(context);
		IValue result = method.interpret(thisContext);
		parentMost.setParentContext(null);
		return result;
	}

	public Identifier getName() {
		return method.getIdentifier();
	}

	public ArgumentList getArguments() {
		return method.getArguments();
	}

	public IType getReturnType() {
		return method.getReturnType();
	}

}

File: Java-CodeServer/src/main/resources/js/prompto/value/ClosureValue.js
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var MethodType = require("../type/MethodType").MethodType;
var Value = require("./Value").Value;

function ClosureValue(context, method) {
    Value.call(this, new MethodType(context, method.name));
    this.context = context;
    this.method = method;
    return this;
}

ClosureValue.prototype = Object.create(Value.prototype);
ClosureValue.prototype.constructor = ClosureValue;

ClosureValue.prototype.interpret = function(context) {
    var thisContext = this.type.context;
    var parentMost = thisContext.getParentMostContext();
    parentMost.setParentContext(context);
    var result = this.method.interpret(thisContext);
    parentMost.setParentContext(null);
    return result;
}

exports.ClosureValue = ClosureValue;
